### PR TITLE
manila: Add HA support for manila-api and manila-scheduler

### PR DESCRIPTION
**Pull request by toabctl**

### CODE REVIEW

File: chef/cookbooks/manila/recipes/common.rb [hunk 1]
```diff
@@ -36,17 +36,23 @@
...
-  bind_host = "0.0.0.0"
+if node[:manila][:ha][:enabled]
+  node[:manila][:api][:bind_host] = my_ipaddress
+  bind_port = node[:manila][:ha][:ports][:api]
 else
-  bind_host = node[:manila][:api][:bind_host]
+  if node[:manila][:api][:bind_open_address]
+    bind_host = "0.0.0.0"
+  else
+    bind_host = node[:manila][:api][:bind_host]
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/common.rb]: I'm a bit confused by your logic with `bind_host` here; isn't `node[:manila][:api][:bind_host]` not set anymore?

Comment (toabctl) [inline chef/cookbooks/manila/recipes/common.rb]: Good catch. Fixed

File: chef/cookbooks/manila/recipes/sql.rb [hunk 2]
```diff
@@ -31,4 +37,32 @@
...
+    node.save
+  end
+  action :nothing
+  subscribes :create, "execute[manila-manage db sync]", :immediately
+end
+
+crowbar_pacemaker_sync_mark "create-manila_database"
+
+# save data so it can be found by search
+node.save
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/sql.rb]: Why is this needed?

File: chef/cookbooks/manila/recipes/controller_ha.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+unless node[:manila][:ha][:enabled]
+  log "HA support for manila is disabled"
+  return
+end
+
+log "HA support for manila is enabled"
+
+haproxy_loadbalancer "manila-api" do
+  address node[:manila][:api][:bind_open_address] ?
+    "0.0.0.0" : node[:manila][:api][:bind_host]
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: I don't believe this is correct. This should be the IP address of haproxy (see chef/cookbooks/glance/libraries/helpers.rb)

File: chef/cookbooks/manila/recipes/controller_ha.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  port node[:manila][:api][:bind_port]
+  # FIXME(toabctl): implement SSL support
+  # use_ssl (node[:manila][:api][:protocol] == "https")
+  use_ssl false
+  servers CrowbarPacemakerHelper.haproxy_servers_for_service(node,
+                                                             "manila",
+                                                             "manila-server",
+                                                             "api")
+  action :nothing
+end.run_action(:create)
```
Comment (aspiers) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Wouldn't it be better to do this after the two `sync_mark`s immediately below?  Are you basing this on another barclamp?  If so, which one?  Because that would either mean that the other barclamp is wrong, or that I'm wrong ;-)

File: chef/cookbooks/manila/recipes/controller_ha.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+pacemaker_clone "cl-#{group_name}" do
+  rsc group_name
+  action [:create, :start]
+  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+end
+
+crowbar_pacemaker_order_only_existing "o-cl-#{group_name}" do
+  # NOTE(toabctl): Add cinder, nova, neutron and glance here? Because these
+  # are needed by the generic driver
```
Comment (vuntz) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Can you find out if we're using the generic driver? If yes, then just change the array depending on the driver.

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 4 above)
Comment (vuntz) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Sorry @aspiers, you're wrong :-)

This resource actually does nothing except setting attribute on the node, so it doesn't matter if it's protected or not.

Comment (toabctl) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: @aspiers it's based on the cinder barclamp code

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 5 above)
Comment (toabctl) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: We can use the generic driver but it's not recommended for production.

File: chef/cookbooks/manila/recipes/sql.rb (hunk 2 above)
Comment (toabctl) [inline chef/cookbooks/manila/recipes/sql.rb]: copied from the cinder barclamp. But it's not needed. We only change ```node``` at line#58  and there is already a ```node.save```

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 3 above)
Comment (toabctl) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Hm. Then it's also wrong in the cinder barclamp (see https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/cinder/recipes/controller_ha.rb#L24 ).

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 4 above)
Comment (aspiers) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: @vuntz And there's no risk of the attribute getting set, the sync_mark failing, and then the attribute taking effect on a later chef-client run even before the `sync_mark` gets passed?

Comment (vuntz) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Nope, the effect is only active in the second phase of chef, so we'd have to pass the sync mark again.

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 3 above)
Comment (vuntz) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: That looks wrong, indeed.

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 4 above)
Comment (aspiers) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Oh right, I didn't notice the `action :nothing` bit.  Even so it's conceptually uglier this way IMHO.  But whatever :)

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 5 above)
Comment (toabctl) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Fixed in https://github.com/crowbar/crowbar-openstack/commit/a5126098ee3293bfdb6d867bff7f223e03c0d906

File: chef/cookbooks/manila/recipes/controller_ha.rb (hunk 3 above)
Comment (toabctl) [inline chef/cookbooks/manila/recipes/controller_ha.rb]: Fixed in https://github.com/crowbar/crowbar-openstack/commit/a5126098ee3293bfdb6d867bff7f223e03c0d906 . I'll prepare a PR for cinder.

File: chef/cookbooks/manila/recipes/sql.rb (hunk 2 above)
Comment (toabctl) [inline chef/cookbooks/manila/recipes/sql.rb]: Fixed in https://github.com/crowbar/crowbar-openstack/commit/a5126098ee3293bfdb6d867bff7f223e03c0d906

File: chef/cookbooks/manila/attributes/default.rb [hunk 6]
```diff
@@ -20,5 +20,15 @@
...
+if %w(redhat centos suse).include? node.platform
+  default[:manila][:ha][:api_ra] = "lsb:openstack-manila-api"
+  default[:manila][:ha][:scheduler_ra] = "lsb:openstack-manila-scheduler"
+else
+  default[:manila][:ha][:api_ra] = "lsb:manila-api"
+  default[:manila][:ha][:scheduler_ra] = "lsb:manila-scheduler"
+end
+default[:manila][:ha][:op][:monitor][:interval] = "10s"
+# Ports to bind to when haproxy is used for the real ports
+default[:manila][:ha][:ports][:api] = 5521
```
Comment (vuntz) [inline chef/cookbooks/manila/attributes/default.rb]: Minor nitpick, but can you bump to 5525 (hopefully not used by any well-known service) so that we can have a few ports after 5520 in case cinder needs another service later on?

**Comment by vuntz** [comment]
Minor nitpick, but +1.

File: chef/cookbooks/manila/attributes/default.rb (hunk 6 above)
Comment (toabctl) [inline chef/cookbooks/manila/attributes/default.rb]: Done in https://github.com/toabctl/crowbar-openstack/commit/975debef18417b3232e49d9fc9b75960a7176ec2

**Comment by dirkmueller** [comment]
+1

**Comment by vuntz** [comment]
+1

**Comment by toabctl** [comment]
Unrelated PR testing failure:
```
10:41:36 [2015-09-21T08:41:34+00:00] INFO: Processing execute[modprobe xfs] action+ complain 73 'Committing the crowbar '\''default'\'' proposal for '\''neutron'\'' failed (1).'
10:41:36 + local ex=73
10:41:36 + shift
10:41:36 + printf 'Error: %s\n' 'Committing the crowbar '\''default'\'' proposal for '\''neutron'\'' failed (1).'
10:41:36 Error: Committing the crowbar 'default' proposal for 'neutron' failed (1).
10:41:36 + [[ 73 = - ]]
10:41:36 + exit 73
```
